docs/rfc/rc-005-event-system.rst

```diff
@@ -0,0 +1,121 @@
+right collection of slow subscription consumers the event system can put
+backpressure on the consensus state machine and message gossiping in the
+network, thereby causing nodes to lag.
+
+Improvements
+~~~~~~~~~~~~
+
+The current system relies on implicit queues built by the buffered channels,
+and though threadsafe, can force all activity within Tendermint to serialize,
+which does not need to happen. Additionally, timeouts for subscription
+consumers related to the implementation of the RPC layer, may complicate the
+use of the system.
+
+References
+~~~~~~~~~~
+
+- Legacy Implementation
+  - `publication of events <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L333-L345>`_ 
+  - `send operation <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L489-L527>`_ 
+  - `send loop <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L381-L402>`_
+- Related RFCs 
+  - `RFC 002: IPC Ecosystem <./rfc-002-ipc-ecosystem.md>`_ 
+  - `RFC 003: Performance Questions <./rfc-003-performance-questions.md>`_ 
+
+Discussion
+----------
+
+Changes to Published Events
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As part of this process, the Tendermint team should do a study of the existing
+event types and ensure that there are viable production use cases for
+subscriptions to all event types. Instinctively it seems plausible that some
+of the events may not be useable outside of tendermint, (e.g. ``TimeoutWait``
+or ``NewRoundStep``) and it might make sense to remove them. Certainly, it
+would be good to make sure that we don't maintain infrastructure for unused or
+un-useful message indefinitely.
+
+Blocking Subscription
+~~~~~~~~~~~~~~~~~~~~~
+
+The blocking subscription mechanism makes it possible to have *send*
+operations into the subscription channel be un-buffered (the event processing
+channel is still buffered.) In the blocking case, events from one subscription
+can block processing that event for other non-blocking subscriptions. The main
+case, it seems for blocking subscriptions is ensuring that a transaction has
+been committed to a block for ``BroadcastTxCommit``. Removing blocking
+subscriptions entirely, and potentially finding another way to implement
+``BroadcastTxCommit``, could lead to important simplifications and
+improvements to throughput without requiring large changes.
+
+Subscription Identification
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Before `#6386 <https://github.com/tendermint/tendermint/pull/6386>`_, all
+subscriptions were identified by the combination of a client ID and a query,
+and with that change, all it became possible to identify subscription's based
+on an ID but compatibility with the legacy identification means that there's a
+good deal of legacy code as well as client side efficiency that could be
+improved. 
+
+Pubsub Changes
+~~~~~~~~~~~~~~
+
+The pubsub core should be implemented in a way that removes the possibility of
```

Comment:
Strong +1 for this high-level idea. I've got some thoughts about how to structure it so we can replace what we're doing now without too much disruption. I can put together a prototype we can play with.

Reply:
I will need some time to write more detail, but here is a high-level synopsis:

1. Split the event system into two layers. 
2. The "inner" layer is a pure time-series with no filtering. Events can only be published to the inner layer, and any component of the node can publish. The inner stream has a (probably configurable) fixed buffer capacity and publishers block when that capacity is reached (see (3)).
3. The "outer" layer is the (only†) obligate subscriber to the inner layer. Its primary job is to drain the inner time-series and index the results. Indexing is the only operation that can push back on the publishers. In addition, the outer layer manages any/all "external" subscribers. A node that doesn't want to expose events can use a null event sink and disable subscriptions from outside the node, and everything will still work.
4. Each "external" subscriber maintains a filter (a function) and a fixed-capacity buffer. When the outer layer delivers an event to an external subscriber, it does not block: The event is either buffered for delivery, or discarded (whether it discards some other event and buffers the new one, or drops the new one, is a separate policy question, but those are basically the two visible outcomes).
5. Each event delivered to the client comes with a "cursor", notionally a timestamp or an index into the underlying time series. This allows a client to specify a point at/after which it wants to "catch up" on events. The amount of event history the node is willing to replay becomes a configurable setting, so clients have enough time to catch up but cannot force the node to maintain unbounded storage.
6. Replaying "old" event state can be handled independently from the event bus: When a client shows up wanting to subscribe to a cursor that is "old", the service can send back (batches of) old event data until it "catches up" then hand off the subscription to the outer layer of the event system to keep going on the "normal" path. This means registration would be a little more complex than it is now, but only for users who care about history.

All of this is very informal right now, but I think it's basically sound, and I think we could do most of the internal plumbing without breaking the existing APIs at first.

† In principle we might want other internal components of the node to be able to subscribe to the inner time series. That would also be fine, and it wouldn't break the model, but I can't think of any important use for that yet.

Reply:
I meant to add: Part of the motivation for decoupling registration/replay from "running" subscribers is to allow us to shed load in a principled way: At the moment, a large number of subscribers can put more or less unlimited memory pressure on the node, even if they are timely about consuming their feed. The node should be able to tell a client at the point of registration "I don't have room for any more subscribers", so it can control both memory usage and concurrency in a predictable way. If there is a need for more capacity, an operator can increase capacity by running one or more "proxy" subscribers that mux external subscription clients onto a smaller set of node subscriptions. We could even write such a proxy—it would need to do some work to compile filters together and so on, but a lot of the code could be shared. The net result is that the system _can_ be scaled up, but the consensus node still has predictable resource footprint and stability characteristics.

Reply:
I think this is more or less what I was thinking of, and think your description is pretty good. I presume that the inner "timeseries" data is readable my subscribers (via their cursors,) after passing through the filter, and only *writeable* by the incoming event stream. 

My inclination is that we should implement this on top of the database layer (e.g. put it on disk rather than memory, and let the storage engine's concurrency control manage most of this,) the "hard" part then becomes deleting old events in an efficient way, as MVCC systems tend to be pretty bad at deletion, but that turns a big collection of hard problems into just one moderately gnarly problem.

Reply:
> I presume that the inner "timeseries" data is readable my subscribers (via their cursors,) after passing through the filter, and only _writeable_ by the incoming event stream.

That's the general idea, yes. The exact point of division between the "inner" and "outer" layers I described isn't crucial—you could imagine, for example, making the "inner" layer responsible for indexing. I described it this way so that there are as few configuration options as possible grafted to the spinal column of the consensus node (e.g., indexing may be disabled entirely without loss of function for the inner layer).

> My inclination is that we should implement this on top of the database layer (e.g. put it on disk rather than memory, and let the storage engine's concurrency control manage most of this,) the "hard" part then becomes deleting old events in an efficient way, as MVCC systems tend to be pretty bad at deletion, but that turns a big collection of hard problems into just one moderately gnarly problem.

Yeah, that's more or less what I was thinking about for indexing as well. The "in memory" part is really just a working buffer. By allowing the indexer to push back the publishers, we can control how much stuff has to land in persistent storage before we "move on". I'm sure we would need to refine the details, but ideally I want to reuse as much of the other infrastructure (e.g., the database) as possible.